sources/pciclient/src/lib.rs

```diff
@@ -0,0 +1,488 @@
+        assert_eq!(
+            list_devices_output[0],
+            ListDevicesOutput {
+                pci_slot: "00:00.0".to_string(),
+                class: "0600".to_string(),
+                vendor: "8086".to_string(),
+                device: "1237".to_string(),
+                program_interface: Some("00".to_string()),
+                subsystem_vendor: Some("1d0f".to_string()),
+                subsystem_device: Some("1237".to_string()),
+                ..Default::default()
+            }
+        );
+        assert_eq!(
+            list_devices_output[1],
+            ListDevicesOutput {
+                pci_slot: "00:01.0".to_string(),
+                class: "0601".to_string(),
+                vendor: "8086".to_string(),
+                device: "7000".to_string(),
+                program_interface: Some("00".to_string()),
+                ..Default::default()
+            }
+        );
+        assert_eq!(
+            list_devices_output[2],
+            ListDevicesOutput {
+                pci_slot: "00:1e.0".to_string(),
+                class: "0302".to_string(),
+                vendor: "10de".to_string(),
+                device: "1eb8".to_string(),
+                program_interface: Some("00".to_string()),
+                revision: Some("a1".to_string()),
+                subsystem_vendor: Some("10de".to_string()),
+                subsystem_device: Some("12a2".to_string()),
+            }
+        );
+    }
+
+    #[test]
+    fn test_list_devices() {
+        let mock_pci_client = MockPciClient {
+            output: vec![r#"00:00.0 "0600" "8086" "1237" -p00 "1d0f" "1237""#.to_string()],
+        };
+        let list_devices_param = ListDevicesParamBuilder::new()
+            .vendor("8086")
+            .device("1237")
+            .class("0600")
+            .program_interface("00")
+            .build();
+        let list_devices_result = call_list_devices(mock_pci_client, list_devices_param);
+        assert!(list_devices_result.is_ok());
+        assert_eq!(
+            list_devices_result.unwrap(),
+            vec![ListDevicesOutput {
+                pci_slot: "00:00.0".to_string(),
+                class: "0600".to_string(),
+                vendor: "8086".to_string(),
+                device: "1237".to_string(),
+                program_interface: Some("00".to_string()),
+                subsystem_vendor: Some("1d0f".to_string()),
+                subsystem_device: Some("1237".to_string()),
+                ..Default::default()
+            }]
+        )
+    }
+
+    #[test]
+    fn test_is_efa_attached() {
+        let mock_pci_client = MockPciClient {
+            // EFA device has class code: 0200 for ethernet controller, vendor 1d0f for amazon, device code efa0.
+            // Below is an actual output from lspci for efa device.
+            output: vec![r#"00:1d.0 "0200" "1d0f" "efa0" -p00 "1d0f" "efa0""#.to_string()],
```

Comment:
Can you add a case that has more than one output to make sure you'll still find it even if there are multiple devices returned, or is this not expected to be the case?